File: Python/myAtoi.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: myAtoi.py
@Function: 字符串转换整数 (atoi)
@Link: https://leetcode-cn.com/problems/string-to-integer-atoi/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-08-16
"""


INT_MAX = 2 ** 31 - 1
INT_MIN = -2 ** 31


class Automaton:
    def __init__(self):
        self.state = 'start'
        self.sign = 1
        self.ans = 0
        self.table = {
            'start': ['start', 'signed', 'in_number', 'end'],
            'signed': ['end', 'end', 'in_number', 'end'],
            'in_number': ['end', 'end', 'in_number', 'end'],
            'end': ['end', 'end', 'end', 'end'],
        }

    def get_col(self, c):
        if c.isspace():
            return 0
        if c == '+' or c == '-':
            return 1
        if c.isdigit():
            return 2
        return 3

    def get(self, c):
        self.state = self.table[self.state][self.get_col(c)]
        if self.state == 'in_number':
            self.ans = self.ans * 10 + int(c)
            self.ans = min(self.ans, INT_MAX) if self.sign == 1 else min(
                self.ans, -INT_MIN)
        elif self.state == 'signed':
            self.sign = 1 if c == '+' else -1


class Solution:
    def myAtoi(self, str: str) -> int:
        automaton = Automaton()
        for c in str:
            automaton.get(c)
        return automaton.sign * automaton.ans


# ----------------------------------
if __name__ == "__main__":
    str = "   -42"
    # str = "4193 with words"

    solution = Solution()
    atoi_number = solution.myAtoi(str)
    print(f"The solution of this problem is {atoi_number}")

File: C++/myAtoi.cpp
/**
 * @File    : myAtoi.cpp
 * @Brief   : 字符串转换整数 (atoi)
 * @Link    : https://leetcode-cn.com/problems/string-to-integer-atoi/
 * @Author  : Wei Li
 * @Date    : 2021-08-16
*/

#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
#include <algorithm>

class Automaton
{
    std::string state = "start";
    std::unordered_map<std::string, std::vector<std::string>> table = {
        {"start", {"start", "signed", "in_number", "end"}},
        {"signed", {"end", "end", "in_number", "end"}},
        {"in_number", {"end", "end", "in_number", "end"}},
        {"end", {"end", "end", "end", "end"}}};

    int get_col(char c)
    {
        if (isspace(c))
            return 0;
        if (c == '+' or c == '-')
            return 1;
        if (isdigit(c))
            return 2;
        return 3;
    }

public:
    int sign = 1;
    long long ans = 0;

    void get(char c)
    {
        state = table[state][get_col(c)];
        if (state == "in_number")
        {
            ans = ans * 10 + c - '0';
            ans = sign == 1 ? std::min(ans, (long long)INT_MAX) : std::min(ans, -(long long)INT_MIN);
        }
        else if (state == "signed")
            sign = c == '+' ? 1 : -1;
    }
};

class Solution
{
public:
    int myAtoi(std::string str)
    {
        Automaton automaton;
        for (char c : str)
            automaton.get(c);
        return automaton.sign * automaton.ans;
    }
};

// -----------------------------
int main(int argc, char **argv)
{
    // std::string str = "   -42";
    std::string str = "4193 with words";

    Solution solution;
    int atoi_number = solution.myAtoi(str);
    std::cout << "The solution of this problem is " << atoi_number << std::endl;

    return 0;
}
